cdap-ui/app/cdap/components/SecureKeys/index.tsx

```diff
@@ -14,166 +14,181 @@
-
-    MySecureKeyApi.list({ namespace }).subscribe((res) => {
-      this.setState({
-        secureKeys: res,
-      });
-    });
-  };
-
-  private handleChange = (property) => {
-    return (e) => {
-      // @ts-ignore
-      this.setState({
-        [property]: e.target.value,
-      });
-    };
+const styles = (): StyleRules => {
+  return {
+    content: {
+      padding: '50px',
+    },
+    loadingBox: {
+      width: '100%',
+      height: '100%',
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+    },
   };
+};
+
+interface ISecureKeysProps extends WithStyles<typeof styles> {}
+
+const SecureKeysView: React.FC<ISecureKeysProps> = ({ classes }) => {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  const { secureKeyStatus, editMode, deleteMode, loading } = state;
+
+  const namespace = getCurrentNamespace();
+
+  // Observe `secureKeyStatus` with `useEffect` and forward the value to `secureKeyStatus$`
+  const secureKeyStatusSubject = React.useRef(new BehaviorSubject(secureKeyStatus));
+  React.useEffect(() => {
+    secureKeyStatusSubject.current.next(secureKeyStatus);
+  }, [secureKeyStatus]);
+  const secureKeyStatus$ = React.useMemo(() => secureKeyStatusSubject.current.asObservable(), [
+    secureKeyStatusSubject,
+  ]);
+
+  // Update the states with incoming secure keys
+  React.useEffect(() => {
+    // Whenever user adds/edits/deletes a secure key, securekeyStatus$ emits a new value.
+    // In such case, re-call MySecureKeyApi.list to reflect the changes
+    const secureKeys$ = secureKeyStatus$
+      .pipe(
+        distinctUntilChanged(),
+        flatMap(() => {
+          return MySecureKeyApi.list({ namespace });
+        })
+      )
+      .publishReplay(1)
+      .refCount();
+
+    const subscription = secureKeys$.subscribe((keys: ISecureKeyState[]) => {
+      dispatch({ type: 'SET_LOADING', loading: true });
+      if (!keys) {
+        return;
+      }
+
+      // Populate the table with matched secure keys
+      dispatch({ type: 'SET_SECURE_KEYS', secureKeys: fromJS(keys) });
+      dispatch({ type: 'SET_LOADING', loading: false });
+    });
 
-  private addKey = () => {
-    const { name, description, data } = this.state;
-
-    if (!name || !data) {
-      return;
-    }
-
-    const namespace = getCurrentNamespace();
-
-    const params = {
-      namespace,
-      key: name,
-    };
-
-    const requestBody = {
-      description,
-      data,
+    return () => {
+      return subscription.unsubscribe();
     };
+  }, []);
 
-    MySecureKeyApi.add(params, requestBody).subscribe(() => {
-      this.setState({
-        name: '',
-        description: '',
-        data: '',
-      });
-      this.fetchSecureKeys();
-    });
+  // Success Alert component always closes after 3000ms.
+  // After a timeout for 3000ms, reset status to make a success Alert component disappear
+  const onSuccessAlertClose = () => {
```

Comment:
what is the difference between the `onSuccessAlertClose` and `onFailureAlertClose`? Seems to be doing the same thing.

Reply:
Resolved